posts/2021/09/array-libraries-interoperability.md

```diff
@@ -0,0 +1,537 @@
+    <i>Array API Demo</i>
+    </a>
+</p>
+
+The [Array API Demo](https://anirudhdagar.ml/array-api-demo/intro.html) walks
+you through the details and processes involved to make an array consumer library
+like SciPy more interoperable with array provider libraries. The demo is built
+keeping two different perspectives in mind, one that of an end user and the other
+of an open-source developer/maintainer looking to incorporate and use Array API
+within their array consumer library.
+
+The demo showcases the 2017 Nobel prize winning work for Physics about
+[LIGO-Virgo detector noise and extraction of transient gravitational-wave signals](https://inspirehep.net/literature/1751757).
+The original tutorial ([collab link](https://colab.research.google.com/github/losc-tutorial/Data_Guide/blob/master/Guide_Notebook.ipynb)) was built using NumPy, SciPy,
+and Matplotlib. 
+But, instead of using NumPy arrays we switch to a PyTorch based
+implementation with minimal changes in the codebase.
+
+Let's dive into the exact formulation and python code that allows this behaviour.
+
+### get_namespace
+
+The demo shows the implementation of a dummy [`get_namespace`](https://anirudhdagar.ml/array-api-demo/GW_Demo_Array_API.html#get-namespace) method which is the
+first function to be called inside any SciPy method. One can see how it works
+below, simply returning the namespace, which can be used later for calling any
+Array API specified methods.
+
+```python
+def get_namespace(*xs):
+    # `xs` contains one or more arrays, or possibly Python scalars (accepting
+    # those is a matter of taste, but doesn't seem unreasonable).
+    namespaces = {
+        x.__array_namespace__() if hasattr(x, '__array_namespace__') else None for x in xs if not isinstance(x, (bool, int, float, complex))
+    }
+
+    if not namespaces:
+        # one could special-case np.ndarray above or use np.asarray here if
+        # older numpy versions need to be supported.
+        # This can happen when lists are sent as an input, for eg. some
+        # SciPy functions coerce lists into ndarrays internally.
+        raise ValueError("Unrecognized array input")
+
+    if len(namespaces) != 1:
+        raise ValueError(f"Multiple namespaces for array inputs: {namespaces}")
+
+    xp, = namespaces
+    if xp is None:
+        raise ValueError("The input is not a supported array type")
+
+    return xp
+
+
+def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+	detrend='constant', return_onesided=True, scaling='density', axis=-1,
+	average='mean'):
+    # Array-API
+    xp = get_namespace(x, y)
+
+    # Library Independent Code
+    # Call Array API specified methods
+    some_internal_calculation = xp.some_func(x)
+    ...
+```
+
+This should be possible with the Array API, for the end-user in a
+future release of SciPy and PyTorch. I'd encourage you to read
+more details on the [demo](https://anirudhdagar.ml/array-api-demo/intro.html)
+page itself.
+
+
```

Comment:
There should be some text here about status of array API implementations, which then makes the transition in the text to your contributions to PyTorch smoother.